_posts/2023-07-09-infra-migration.md

```diff
@@ -0,0 +1,332 @@
+
+Using Nomad hasn't been a flawless experience.
+As we want to be as cheap as possible (which happens when you depend on donations), we are using AWS's T4g EC2-instances.
+These machines are ARM64.
+And Nomad had some open issues in those regards.
+
+But, nothing that can't be fixed, and so far they have been great accepting contributions.
+That is the lovely thing about a company that makes their products Open Source: others can contribute to make it better.
+So by now I fixed the two problems that were holding me back from using it on ARM64, and I hope others enjoy those fixes too.
+
+With Nomad, I have full insight what a service is doing.
+How much memory, how much CPU, how often they crashed, etc.
+The one small thing I am still missing: I cannot see which service uses how much bandwidth.
+But I plan on addressing that soon ... I have some ideas.
+
+And I haven't setup log collection yet; I intend to use [Loki](https://grafana.com/oss/loki/) for this, as their Free offering is excellent for our goal.
+With Loki, we can look back in logs of services, even if the Nomad client crashed and died.
+
+A short side-step: Nomad runs on an AWS Autoscaling Group.
+We automated this completely, which means we can scale up the amount of nodes and scale down easily.
+This means that updating to a new Nomad or OS is not some manual work: you just scale up the cluster, and slowly scale it down again.
+That way, new machines are created with the latest version, and the old ones are slowly shut down.
+Far far easier than trying to manually update machines; and much less likely to break.
+
+## GitHub Workflows
+
+The one thing not talked about above, is that we also really badly needed to address our [GitHub Workflows](https://docs.github.com/en/actions/using-workflows).
+GitHub Workflows automate our repositories, but were written in a time where [Reusing Workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) didn't exist yet.
+
+The problem was maintaining the GitHub Workflows; it would require changes in ~20 repositories.
+This was very time consuming and error-prone.
+
+By switching to Reusing Workflows, it means we have a [single repository](https://github.com/OpenTTD/actions) which hosts all our workflows.
+By changing things in a single place, all repositories will be updated.
+
+This will make our life so much easier.
+
+While at it, we also noticed that making releases of those repositories to get things to production wasn't really working, and was an annoying task.
+So we decided to also change that: if you now merge a Pull Request, it is deployed to production.
+No delay.
+
+And to compensate for the missing staging, you can now label a Pull Request with a preview label in any of the repositories, and it will deploy this to a preview site.
+Here you can review the changes, and if you like it, merge the Pull Request.
```

Comment:
```suggestion
And to compensate for the missing staging, we can now label a Pull Request with a preview label in any of the repositories, and it will deploy this to a preview site.
Here we can review the changes, and if we like it, merge the Pull Request.
```